src/main/java/com/conveyal/datatools/manager/jobs/ProcessSingleFeedJob.java

```diff
@@ -42,20 +66,110 @@ public String getFeedVersionId () {
+            for (ZipTransformation transformation : zipTransformations) {
+                ArbitraryTransformJob zipTransform = new ArbitraryTransformJob(owner, feedVersion.retrieveGtfsFile(), transformation);
+                // Run transform job in line so we can monitor the error status before load/validate begins.
+                zipTransform.run();
+                // Short circuit the feed load/validate if a pre-load transform fails.
+                if (zipTransform.status.error) return;
+            }
+        }
 
         // First, load the feed into database. During this stage, the GTFS file will be uploaded to S3 (and deleted locally).
         addNextJob(new LoadFeedJob(feedVersion, owner, isNewVersion));
 
         // Next, validate the feed.
         addNextJob(new ValidateFeedJob(feedVersion, owner, isNewVersion));
 
+        // We only need to snapshot the feed if there are transformations at the database level. In the case that there
+        // are, the snapshot namespace will be the target of these modifications. If we were to apply the modifications
+        // to the original feed version's namespace, the namespace and the zip file would become out of sync, which
+        // violates a core expectation for this application: that each zip file represents exactly its database
+        // representation.
+        boolean shouldSnapshot = shouldTransform && rules.hasTransformationsOfType(feedVersion, DbTransformation.class);
+        if (shouldSnapshot) {
+            Snapshot snapshot = new Snapshot("Transform of " + feedVersion.name, feedVersion);
+            addNextJob(new CreateSnapshotJob(owner, snapshot));
+            // Apply post-load transformations to snapshotted feed version. Post-load transformations will modify only the
+            // snapshot (not the original feed version's namespace), so the snapshot must be published (or loaded into the
+            // editor) in order to see the results.
+            // FIXME need to add snapshot namespace...
+            for (DbTransformation transformation : rules.getActiveTransformations(feedVersion, DbTransformation.class)) {
+                addNextJob(new ArbitraryTransformJob(owner, snapshot.id, transformation));
+            }
+            // If the user has selected to create a new version from the resulting snapshot, do so here.
+            if (rules.createNewVersion) {
+                addNextJob(new CreateFeedVersionFromSnapshotJob(feedSource, snapshot, owner));
+            }
+        }
+
+        if (shouldClone) {
+            // If it was determined that the final version should be cloned (because of feed transformations that
+            // must operate on a cloned version), handle that here. The cloned version will then apply the
+            // transformations in its process feed job.
+            try {
+                // Create a new version for the clone.
+                FeedVersion newFeedVersion = new FeedVersion(feedSource, VERSION_CLONE);
```

Comment:
So how many FeedVersions will get created here? This seems like it would have the potential to clutter things up. Let's say a few transformations exist on a feed that is fetched nightly. Instead of just one single FeedVersion being created, there could be multiple... along with additional snapshots. Seems like we may want to just drop this expectation that a FeedVersion exactly represents an incoming GTFS file if there are transforms that get applied. We should still save the original GTFS and perhaps the pre-processed input GTFS, but having a bunch of FeedVersions and intermediate snapshots created seems like we're trying to use existing tools instead of making a proper new tool.

Reply:
Instead of creating all of these feed versions and snapshots, after discussion we will just try to create one to avoid confusion for users. We can just show a transformation summary for any version that was modified.